src/support/sparse_square_matrix.h

```diff
@@ -0,0 +1,63 @@
+/*
+ * Copyright 2022 WebAssembly Community Group participants
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+// A helper type for potentially sparse N*N matrix container.
+
+#pragma once
+
+#include <stdint.h>
+#include <unordered_map>
+#include <vector>
+
+template<typename Ty> class sparse_square_matrix {
+  std::vector<Ty> denseStorage;
+  std::unordered_map<uint64_t, Ty> sparseStorage;
+  uint32_t N;
+
+public:
+  sparse_square_matrix() : N(0) {}
```

Comment:
```suggestion
  sparse_square_matrix() : N(0) {
    assert(N > 0);
  }
```
(just to avoid corner cases with ambiguity around dense or sparse storage being used)

Reply:
Sorry, probably some typo here, or? The assert would always trigger here, as default initialization initializes N to zero, and 0 > 0 would be false.

Reply:
Oh now I think I understand your intent.

If the matrix is of 0*0 size, then it does not matter if it is categorized to use dense or sparse storage, since there won't be any elements stored in it, and it won't be valid to call set() or get() with any index. Added assert()s to the get() and set() functions to check that the indices that the functions are called with will be within bounds.